sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -304,8 +503,195 @@ class RapidsBufferCatalog extends AutoCloseable with Arm {
     bufferMap.compute(buffer.id, updater)
   }
 
-  /** Remove a buffer ID from the catalog at the specified storage tier. */
-  def removeBufferTier(id: RapidsBufferId, tier: StorageTier): Unit = {
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  def synchronousSpill(store: RapidsBufferStore, targetTotalSize: Long): Option[Long] =
+    synchronousSpill(store, targetTotalSize, Cuda.DEFAULT_STREAM)
+
+  /**
+   * Free memory in `store` by spilling buffers to the spill store synchronously.
+   * @param store store to spill from
+   * @param targetTotalSize maximum total size of this store after spilling completes
+   * @param stream CUDA stream to use or null for default stream
+   * @return optionally number of bytes that were spilled, or None if this called
+   *         made no attempt to spill due to a detected spill race
+   */
+  private def synchronousSpill(
```

Comment:
Again, why are these separate? The private version that takes a Cuda.Stream is not called anywhere, except just above here. Also is there any time that we would not want to use the default stream?

Reply:
I can merge the methods, good eye.

On the default stream: I think default stream makes sense for PTDS, but for non-PTDS. We are not prepared to handle that optimally.

- The expectation is that buffers added to the spill store are synchronized w.r.t the stream that is doing the adding (`addBuffer` winds up synchronizing, unless it's from the shuffle code which guarantees it synchronizes buffers before calling `catalog.addBuffer`). 

- On spill, the PTDS stream is doing the copying, so it is the one responsible for synchronization, and default stream seems like the right choice for PTDS (but not for non-PTDS, this code is going to over synchronize on spill).

- On un-spill, or the memcpy-and-throwaway that happen if unspill=false, we are using the `cudaMemcpy` on the default stream (forced, the stream is not passed to those apis from calling code) and we are synchronizing on this default stream. Again, this is mostly OK because the result of the memcpy is after that the buffer is re-added to the store as a `RapidsDeviceMemoryBuffer` or returned to the user as a pointer to `DeviceMemoryBuffer`.